include/log.h

```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#include <ostream>
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <unordered_map>
+#include <variant>
+
+// TODO: Make this configurable through CMake?
+#define SPDLOG_ACTIVE_LEVEL SPDLOG_LEVEL_TRACE
+#include <spdlog/spdlog.h>
+
+#include <spdlog/fmt/ostr.h>
+#include <spdlog/sinks/stdout_color_sinks.h>
+
+#include "print_utils.h"
+
+#define CELERITY_LOG_SET_SCOPED_CTX(ctx) CELERITY_DETAIL_LOG_SET_SCOPED_CTX(ctx)
+
+#define CELERITY_TRACE(...) SPDLOG_TRACE("{}{}", *celerity::detail::active_log_ctx, fmt::format(__VA_ARGS__))
+#define CELERITY_DEBUG(...) SPDLOG_DEBUG("{}{}", *celerity::detail::active_log_ctx, fmt::format(__VA_ARGS__))
+#define CELERITY_INFO(...) SPDLOG_INFO("{}{}", *celerity::detail::active_log_ctx, fmt::format(__VA_ARGS__))
+#define CELERITY_WARN(...) SPDLOG_WARN("{}{}", *celerity::detail::active_log_ctx, fmt::format(__VA_ARGS__))
+#define CELERITY_ERROR(...) SPDLOG_ERROR("{}{}", *celerity::detail::active_log_ctx, fmt::format(__VA_ARGS__))
+#define CELERITY_CRITICAL(...) SPDLOG_CRITICAL("{}{}", *celerity::detail::active_log_ctx, fmt::format(__VA_ARGS__))
+
+namespace celerity {
+namespace detail {
+
+	using log_map = std::unordered_map<std::string, std::variant<std::string_view, size_t>>;
+	using log_level = spdlog::level::level_enum;
+
+	struct log_context {
+		std::string value;
+		log_context() = default;
+		explicit log_context(const log_map& values) { value = fmt::format("[{}] ", values); }
+	};
+
+	inline const std::string null_log_ctx;
+	inline thread_local const std::string* active_log_ctx = &null_log_ctx;
+
+	struct log_ctx_setter {
+		log_ctx_setter(log_context& ctx) { celerity::detail::active_log_ctx = &ctx.value; }
+		~log_ctx_setter() { celerity::detail::active_log_ctx = &celerity::detail::null_log_ctx; }
+	};
+
+#define CELERITY_DETAIL_LOG_SET_SCOPED_CTX(ctx)                                                                                                                \
+	log_ctx_setter _set_log_ctx_##__COUNTER__ { ctx }
+
+} // namespace detail
+} // namespace celerity
+
+template <>
+struct fmt::formatter<celerity::detail::log_map> {
+	constexpr auto parse(format_parse_context& ctx) { return ctx.begin(); }
+
+	template <typename FormatContext>
+	auto format(const celerity::detail::log_map& map, FormatContext& ctx) {
+		auto&& out = ctx.out();
+		int i = 0;
+		for(const auto& [k, v] : map) {
+			if(i++ > 0) { fmt::format_to(out, ", "); }
+			fmt::format_to(out, "{}=", k);
+			if(std::holds_alternative<std::string_view>(v)) {
```

Comment:
Is there any benefit in using the variant here instead of making the `log_map` a `map<string, string>` and requiring the user to call `fmt::format` themselves or adding a templated insertion function to `log_map`? `string_view` is a potential footgun because it is implicitly converted from a string but requires the user to keep the backing string alive.

If this is about performance, we should allow disabling insertion into the `log_map` via macros just like we do for the rest of the logging API.

If we keep the variant, I suggest simplifying the if-statement (and the preceding `format_to`) to
```c++
std::visit([&](auto &variant) { fmt::format_to(out, "{}={}", k, variant); }, v);
```

Reply:
It was a performance consideration, although creating the `unordered_map` was likely much more expensive anyway. I agree on the footgunnyness of `string_view`. I've gotten rid of the map altogether and replaced it with a `std::tuple` that can hold arbitrary types (as long as `fmt` knows how to format them).

I didn't know of `std::visit` btw, pretty neat.